src/negotiation/order.ts

```diff
@@ -16,8 +16,93 @@ export interface OrderAsset {
   nominalAmount: string;
 }
 
+export interface TakerCriteria {
+  buy: TakerCriteriaAsset;
+  sell: TakerCriteriaAsset;
+}
+
+export interface TakerCriteriaAsset {
+  ledger: string;
+  asset: string;
+  minNominalAmount?: string;
+  maxNominalAmount?: string;
+}
+
+export class Order {
+  constructor(
+    public readonly orderParams: OrderParams,
+    public readonly criteria: TakerCriteria,
+    public readonly takeOrder: (
+      orderParams: OrderParams
+    ) => Promise<Swap | undefined>
+  ) {}
+
+  public matches(): boolean {
+    return this.assetsMatch();
+  }
+
+  public isValid(): boolean {
+    if (
+      !(
+        this.orderParams.ask.ledger &&
+        this.orderParams.ask.asset &&
+        this.orderParams.ask.nominalAmount &&
+        this.orderParams.bid.ledger &&
+        this.orderParams.bid.asset &&
+        this.orderParams.bid.nominalAmount &&
+        this.orderParams.validUntil &&
+        this.orderParams.id
+      )
+    ) {
+      return false;
+    }
+
+    const askAmount = new BigNumber(this.orderParams.ask.nominalAmount, 10);
+    const bidAmount = new BigNumber(this.orderParams.bid.nominalAmount, 10);
+
+    return !askAmount.isNaN() && !bidAmount.isNaN();
+  }
+
+  public async take(): Promise<Swap | undefined> {
+    if (this.matches() && this.isValid()) {
```

Comment:
I would turn the match around and check for validity first. Also, you can call `assetsMatch` directly.

```suggestion
    if (this.isValid() && this.assetsMatch()) {
```